### ISSUE TITLE
Decorators confuse E1120 analysis

### ISSUE THREAD

**Issue by pylint-bot** [PyCQA/pylint#259]
Originally reported by: **the mulhern (BitBucket: [the_mulhern](http://bitbucket.org/the_mulhern))**

---

Here's the example:

```
#!python


def decorator(f):

    def new_func(junk=None):
        f(2, junk=junk)

    return new_func

@decorator
def junk1(param, junk=None):
    print("%s" % junk)

def main():
    junk1(junk=2)
```

The analysis reports
"No value passed for parameter 'param' in function call (no-value-for-parameter)"
This is incorrect, since the actual function being called is the new_func manufactured
by the decorator, and that function does not require the parameter 'param'.

You can work around this by giving param a default value, (which will never be used).

---
- Bitbucket: https://bitbucket.org/logilab/pylint/issue/259

**Comment by deckar01** [comment]
Comments moved here from #207.

---

Here is a minimal example to reproduce the issue:

```py
def helper(fun):
    def wrapper(b):
        return fun('helper', b)
    return wrapper

@helper
def simple(a, b):
    print a, b

simple('test')
```

> E: 10, 0: No value for argument 'b' in function call (no-value-for-parameter)

---

Here is the same example with the decorator replaced with method composition:

```py
def helper(fun):
    def wrapper(b):
        return fun('helper', b)
    return wrapper

def hard(a, b):
    print a, b

simple = helper(hard)

simple('test')
```

> No errors

---

To prove that pylint is not simply skipping the signature checking for method composition:

```py
def helper(fun):
    def wrapper(b):
        return fun('helper', b)
    return wrapper

def hard(a, b):
    print a, b

simple = helper(hard)

simple('helper?', 'test')
```

> E: 11, 0: Too many positional arguments for function call (too-many-function-args)

---

The issue is not that pylint is incapable of inferring the signature of generated methods, but rather that pylint lacks the instruction to treat decorators as the composed functions they are.

---

This issue seems to stem from the core AST module leaving it up to the consumer to resolve the signature produced by applying the `decorator_list`.

It should be sufficient to detect when a function node has decorators and substitute it with the return value(s?) of the first decorator.

I think it would be generally useful to determine the resulting signature in asteroid. Maybe an attribute like `external_args`?

**Comment by eight04** [comment]
Thanks @deckar01. Converting the decorator to simple function call did work.

**Comment by iddan** [comment]
Any change with this?

**Comment by davidjlloyd** [comment]
This is still broken as of pylint 1.7.1, astroid 1.5.3.

**Comment by maozguttman** [comment]
What about following code (based on [deckar01](https://github.com/PyCQA/pylint/issues/259#issuecomment-267671718)'s above example) which gives false positive too-many-function-args error:
```python
def helper(fun):
    def wrapper(a, b, *args, **kwargs):
        print "helper wrapper: a={}".format(a)
        return fun(b, *args, **kwargs)
    return wrapper

@helper
def simple(b, c):
    print "simple: b={} c={}".format(b, c)

simple('argentina', 'brazil', 'china')
```
Its output is:
```
helper wrapper: a=argentina
simple: b=brazil c=china
```
And pylint  false positive error is:
```
E: 11, 0: Too many positional arguments for function call (too-many-function-args)
```
Thanks,
Maoz

**Comment by Niedzwiedzw** [comment]
Is this still broken?

**Comment by PCManticore** [comment]
Hey folks, this issue is not fixed yet, as it is still opened. I can't promise when it's going to be fixed though.

**Comment by isaacnorman82** [comment]
This is marked as closed but I still see this issue in pylint 2.5.3.  Am I missing something?

**Comment by tmattha** [comment]
I am still having a false-positive with hypothesis.strategies.composite in pylint 2.5.3.
Error message: No value for argument 'draw' in function call
see also: https://hypothesis.readthedocs.io/en/latest/data.html#composite-strategies

**Comment by bionicles** [comment]
Same issue with 

```
def curry(fun):
    return partial(partial, fun)
```

**Comment by UncleOwen** [comment]
@isaacnorman82 @tmattha @bionicles 
Have a look at https://github.com/PyCQA/pylint/pull/2926.

**Comment by bionicles** [comment]
@UncleOwen how can we activate it?

**Comment by bionicles** [comment]
`path/to/repo/pylint.rc`
```
[TYPECHECK]
signature-mutators=toolz.curry
```

**Comment by bionicles** [comment]
still not working... not sure what i did before but now it doesn't work and drives me nuts

**Comment by bionicles** [comment]
this made me rage, but i finally fixed it, so here's a fix for the next frustrated person

src.ramda
```
def curry(arity):
    """curries a function with a pre-determined number of arguments"""

    def currier(func):
        """curries a function with a pre-determined number of arguments"""

        def curried(*args, **kwargs):
            if len(args) + len(kwargs) >= arity:
                return func(*args, **kwargs)
            new_func = lambda *args2, **kwargs2: curried(
                *(args + args2), **dict(kwargs, **kwargs2)
            )
            new_func._bound_args = args  # pylint: disable=protected-access
            return new_func

        functools.update_wrapper(curried, func)
        curried.__doc__ = func.__doc__
        return curried

    return currier

curry2 = curry(2)
curry3 = curry(3)
curry4 = curry(4)
```

<repo>/pylintrc
```
[MASTER]
ignore=archive,data,scripts,art
jobs=2

[TYPECHECK]
signature-mutators=src.ramda.curry,src.ramda.curry2,src.ramda.curry3,src.ramda.curry4
```

**Comment by hippo91** [comment]
This issue become too messy. I lock it. If you are facing a problem that looks similar please open a new issue that refers this one if necessary.
